### PR TITLE
filter volume list/info on basis of metadata keys and values

### PR DESCRIPTION
**Pull request by vpandey-RH**
Updates [#462](https://github.com/gluster/glusterd2/issues/462).

Add a filter for volumes when doing volume info/list to search through the metadata based on key or value or key + value

Signed-off-by: Vishal Pandey <vpandey@redhat.com>

### CODE REVIEW

File: pkg/restclient/volume.go [hunk 1]
```diff
@@ -3,7 +3,9 @@ package restclient
 import (
 	"fmt"
 	"net/http"
+	"net/url"
 
+	"github.com/gluster/glusterd2/glusterd2/volume"
```
Comment (prashanthpai) [inline pkg/restclient/volume.go]: Restclient package should not import any of the core packages except `api` packages.

File: glusterd2/volume/store-utils.go [hunk 2]
```diff
@@ -91,17 +102,38 @@ func GetVolumesList() (map[string]uuid.UUID, error) {
 	return volumes, nil
 }
 
+// GetFilterType return the filter type for volume list/info
+func GetFilterType(filterParams map[string]string) Filter {
```
Comment (prashanthpai) [inline glusterd2/volume/store-utils.go]: Unexport this function, do not import this in restclient.

File: e2e/volume_ops_test.go [hunk 3]
```diff
@@ -100,6 +100,9 @@ func testVolumeCreate(t *testing.T) {
 				},
 			},
 		},
+		Metadata: map[string]string{
+			"owner": "gd2test",
```
Comment (prashanthpai) [inline e2e/volume_ops_test.go]: Add a value here.

File: e2e/volume_ops_test.go [hunk 4]
```diff
@@ -148,8 +151,51 @@ func testVolumeStop(t *testing.T) {
 
 func testVolumeList(t *testing.T) {
 	r := require.New(t)
+	filterParams := map[string]string{
+		"key":   "owner",
+		"value": "gd2test",
+	}
+	volumes, err := client.Volumes("", filterParams)
+	r.Nil(err)
+	r.Len(volumes, 1)
```
Comment (prashanthpai) [inline e2e/volume_ops_test.go]: When both key and value is present, return only those volumes where both matches. This doesn't seem to do that as per the code above in `testVolumeCreate` where you've given only key.

File: e2e/volume_ops_test.go [hunk 5]
```diff
@@ -148,8 +151,51 @@ func testVolumeStop(t *testing.T) {
...
+
+	filterParams = map[string]string{
+		"key": "owner",
+	}
+	volumes, err = client.Volumes("", filterParams)
+	r.Nil(err)
+	r.Len(volumes, 1)
+
```
Comment (prashanthpai) [inline e2e/volume_ops_test.go]: Write some comment here to separate match vs no-match cases.

* key match
* value match
* key and value match

You can create an array of `map[string]string` values and loop over the matching path.

```go
for _, filter := range matchingQueries {
    volumes, err = client.Volumes("", filter)
    r.Nil(err)
    r.Len(volumes, 1)
}
```

Do the same for non-matching cases:

```go
for _, filter := range NonMatchingQueries {
    volumes, err = client.Volumes("", filter)
    r.Nil(err)
    r.Len(volumes, 0)
}
```

File: glusterd2/volume/store-utils.go [hunk 6]
```diff
@@ -16,6 +16,17 @@ const (
 	volumePrefix string = "volumes/"
 )
 
+// Filter is a filter type
+type Filter uint32
+
+// GetVolumes Filter Types
+const (
+	KeyAndValueNotFound Filter = iota
```
Comment (prashanthpai) [inline glusterd2/volume/store-utils.go]: The `Found` suffix might be misleading. You can remove

File: glusterd2/volume/store-utils.go [hunk 7]
```diff
@@ -16,6 +16,17 @@ const (
 	volumePrefix string = "volumes/"
 )
 
+// Filter is a filter type
+type Filter uint32
```
Comment (prashanthpai) [inline glusterd2/volume/store-utils.go]: Do not export this type. You can also name this `metadataFilter`

File: glusterd2/volume/store-utils.go [hunk 8]
```diff
@@ -111,8 +143,26 @@ func GetVolumes() ([]*Volinfo, error) {
...
+				volumes = append(volumes, &vol)
+			}
+		case OnlyValueFound:
+			for _, value := range vol.Metadata {
+				if value == filterParams[0]["value"] {
+					volumes = append(volumes, &vol)
+				}
+			}
+		case KeyAndValueFound:
+			if value, keyFound := vol.Metadata[filterParams[0]["key"]]; value == filterParams[0]["value"] && keyFound {
```
Comment (prashanthpai) [inline glusterd2/volume/store-utils.go]: For slightly better readability:

```go
if value, keyFound := vol.Metadata[filterParams[0]["key"]]; keyFound
    if  value == filterParams[0]["value"] {
        ...
    }
}
```

File: glusterd2/volume/store-utils.go [hunk 9]
```diff
@@ -111,8 +143,26 @@ func GetVolumes() ([]*Volinfo, error) {
 			}).Error("Failed to unmarshal volume")
 			continue
 		}
+		switch filterType {
+
+		case KeyAndValueNotFound:
```
Comment (prashanthpai) [inline glusterd2/volume/store-utils.go]: You don't have to handle this specially, this can be under `default` case.

**Comment by prashanthpai** [comment]
retest this please

**Comment by vpandey-RH** [comment]
@prashanthpai  Is there anywhere I can mention about CLI flags also ?

**Comment by prashanthpai** [comment]
> @prashanthpai Is there anywhere I can mention about CLI flags also ?

Don't worry about those. They will be seen in `glustercli --help`

**Pull request merged**